### ISSUE TITLE
prediction fails on custom vectors

### ISSUE THREAD

**Issue by bkamins** [JuliaStats/GLM.jl#442]
In https://github.com/JuliaStats/GLM.jl/blob/96f6eeba0f93c1271982e75575e303b354a2b412/src/linpred.jl#L129 (and related lines) we require `Vector{T}` in `delbeta!`. However, in general the user might pass some other type of `AbstractVector` and prediction fails.

I am not too much into the GLM.jl package to know the best approach to fix it (the safest one would be to make the signatures more flexible force conversion to `Vector`).

CC @quinnj + @nalimilan - I have caught it by checking what things the usage of `ChainedVector` can break, but this time the issue is on GLM.jl side.

**Comment by nalimilan** [comment]
Given that this code is relatively low-level and relies on `BlasReal`, it seems safest to convert to `Vector{T}`. `ldiv!` for `Cholesky` seems to support only `StridedArray` objects anyway.

**Comment by nalimilan** [comment]
That said, the vector is only used in `mul!(p.delbeta, transpose(p.X), r)`, so as long as `mul!` works it's probably fine to allow any kind of vector. If it works maybe it's OK to widen the signature.

**Comment by bkamins** [comment]
I meant to widen the signature and internally do `convert(Vector{T}, whatever_was_passed)`, so for `Vector` it would be a no-op.

**Comment by nalimilan** [comment]
Yeah that's what I understood. But in my second comment I wondered whether just widening the signature would be enough.

**Comment by bkamins** [comment]
I have checked that `mul!` is fully generic, so I have opened https://github.com/JuliaStats/GLM.jl/pull/443.

**Comment by andreasnoack** [comment]
Do you have a reproducer for this? It would be useful to see the exact code path that triggers this. As mentioned in https://github.com/JuliaStats/GLM.jl/pull/443#issuecomment-884476200, I'd think that it's better to initialize `r` differently to ensure that it's a `Vector`. It will be much easier to ensure reasonable performance with a more restrictive signature.

**Comment by bkamins** [comment]
I have included the reproducers in tests of #443. Here they are repeated (notice that in #443 I have spotted that also `cholpred` requires fixing not only `delbeta!`):
```
julia> using GLM, SparseArrays

julia> X = sprand(5, 3, 0.5)
5×3 SparseMatrixCSC{Float64, Int64} with 7 stored entries:
 0.370538  0.226452   ⋅
 0.20658    ⋅         ⋅
  ⋅         ⋅        0.0668819
 0.316714   ⋅         ⋅
 0.465847   ⋅        0.853667

julia> y = rand(5)
5-element Vector{Float64}:
 0.8026875897870902
 0.7832594240669271
 0.8125885913468363
 0.6781017823878368
 0.05339301252861639

julia> fit(LinearModel, X, y)
ERROR: MethodError: no method matching cholpred(::SparseMatrixCSC{Float64, Int64}, ::Bool)

julia> fit(LinearModel, X, sparse(y))
ERROR: MethodError: no method matching cholpred(::SparseMatrixCSC{Float64, Int64}, ::Bool)

julia> fit(LinearModel, Matrix(X), sparse(y))
ERROR: MethodError: no method matching delbeta!(::GLM.DensePredChol{Float64, LinearAlgebra.CholeskyPivoted{Float64, Matrix{Float64}}}, ::SparseVector{Float64, Int64})
```
